### PR TITLE
Add snap package manager support to data provider

### PR DESCRIPTION
**Pull request by JorgeMarinoDev**
|Wazuh version|Component|Install type|Install method|Platform|
|---|---|---|---|---|
| feature | syscollector | Agent | Packages/Sources | linux |

## Related Issue
[15429](https://github.com/wazuh/wazuh/issues/15429)

## Description
In order to get the information from snapcraft (ubuntu's package manager installed by default) it is necessary to make a query through the api that this service exposes.

This was implemented using the [Wazuh Http Request Module](https://github.com/wazuh/wazuh-http-request) to connect to standard unix socket exposed by snapd daemon. This daemon has REST V2 [API](https://snapcraft.io/docs/snapd-api#heading--snaps)

## Tests
Added parseSnap* unit tests. 
image

Update:
In order to facilitate obtaining the `http-request` submodule in areas where obtaining the git submodulos may be limited, the facility to obtain this sub-project when downloading the external dependencies is implemented as an alternative, the condition is that the folder `src/shared_modules/http-request` is empty at the time of `make deps`, at the same time it is allowed to select the branch from where the `http-request` is obtained, using the variable `HTTP_REQUEST_BRANCH`, by default it is obtained from main. e.g.

`make deps HTTP_REQUEST_BRANCH=<branch>`

## snap packages recognition evidence
image

### CODE REVIEW

File: src/data_provider/src/packages/packageLinuxParserHelper.h [hunk 1]
```diff
@@ -133,6 +135,113 @@ namespace PackageLinuxHelper
...
+                {
+                    size = std::stoi( stringData );
+                }
+            }
+        }
+
+        ret["name"]             = name;
+        ret["location"]         = "/snap/" + name;
+        ret["version"]          = version;
+        ret["vendor"]           = vendor;
```
Comment (pereyra-m) [inline src/data_provider/src/packages/packageLinuxParserHelper.h:229]: In case some of the values weren't found, we should insert `UNKNOWN_VALUE`. This section assumes that there is always a vendor, description, etc. field.

File: src/data_provider/tests/sysInfoPackagesLinuxHelper/sysInfoPackagesLinuxHelper_test.cpp [hunk 2]
```diff
@@ -433,3 +433,49 @@ TEST_F(SysInfoPackagesLinuxHelperTest, parsePacmanInformationNull)
...
+{
+    std::string jData = "{\"id\":\"rw36mkAjdIKl13dzfwyxP87cejpyIcct\",\"title\":\"gnome-3-38-2004\",\"summary\":\"Shared GNOME 3.38 Ubuntu stack\",\"description\":\"This snap includes a GNOME 3.38 stack (the base libraries and desktop \\nintegration components) and shares it through the content interface. \\n\",\"icon\":\"/v2/icons/gnome-3-38-2004/icon\",\"installed-size\":363151360,\"name\":\"gnome-3-38-2004\",\"publisher\":{\"id\":\"canonical\",\"username\":\"canonical\",\"display-name\":\"Canonical\",\"validation\":\"verified\"},\"developer\":\"canonical\",\"status\":\"active\",\"type\":\"app\",\"base\":\"core20\",\"version\":\"0+git.6f39565\",\"channel\":\"latest/stable\",\"tracking-channel\":\"latest/stable/ubuntu-22.04\",\"ignore-validation\":false,\"revision\":\"119\",\"confinement\":\"strict\",\"private\":false,\"devmode\":false,\"jailmode\":false,\"contact\":\"\",\"mounted-from\":\"/var/lib/snapd/snaps/gnome-3-38-2004_119.snap\",\"media\":[{\"type\":\"icon\",\"url\":\"https://dashboard.snapcraft.io/site_media/appmedia/2021/01/icon_FvbmexL.png\"}],\"install-date\":\"2022-11-23T20:33:59.025662696-03:00\"}";
```
Comment (pereyra-m) [inline src/data_provider/tests/sysInfoPackagesLinuxHelper/sysInfoPackagesLinuxHelper_test.cpp:439]: This line is too long, please split it. 
Repeat for all the lines below.

File: src/data_provider/tests/sysInfoPackagesLinuxHelper/sysInfoPackagesLinuxHelper_test.cpp [hunk 3]
```diff
@@ -433,3 +433,49 @@ TEST_F(SysInfoPackagesLinuxHelperTest, parsePacmanInformationNull)
...
+    nlohmann::json jsonInput = nlohmann::json::parse(jData);
```
Comment (pereyra-m) [inline src/data_provider/tests/sysInfoPackagesLinuxHelper/sysInfoPackagesLinuxHelper_test.cpp:440]: Just a suggestion: instead of creating a string and parsing it, you could generate a `nlohmann::json` literal. See https://github.com/nlohmann/json#creating-json-objects-from-json-literals

File: src/data_provider/tests/sysInfoPackagesLinuxHelper/sysInfoPackagesLinuxHelper_test.cpp [hunk 4]
```diff
@@ -433,3 +433,49 @@ TEST_F(SysInfoPackagesLinuxHelperTest, parsePacmanInformationNull)
...
+
+    const auto& jsPackageInfo { PackageLinuxHelper::parseSnap(jsonInput) };
+    EXPECT_FALSE(jsPackageInfo.empty());
+    EXPECT_EQ(363151360, jsPackageInfo["size"]);
+}
```
Comment (pereyra-m) [inline src/data_provider/tests/sysInfoPackagesLinuxHelper/sysInfoPackagesLinuxHelper_test.cpp:481]: Could you add some more tests? Like: a JSON with a `NULL` field, an empty JSON, etc.

**Review by pereyra-m — COMMENTED** [review]
GJ!

File: src/data_provider/src/packages/packageLinuxParserHelper.h [hunk 5]
```diff
@@ -133,6 +135,113 @@ namespace PackageLinuxHelper
...
+        std::string priority;
+        std::string multiarch;
+        std::string source;
+        std::string format;
+        int         size = 0;
+
+        bool        hasName    = false;
+        bool        hasVersion = false;
+
+        // TODO: Dismiss record if name or version are not present.
```
[the remark this answers is not in the document]
Comment (MiguelazoDS) [reply src/data_provider/src/packages/packageLinuxParserHelper.h:160]: I think more than updating the comment, the code must be changed to implement the feature. We can not accept a package without a name or version.

File: src/Makefile [hunk 6]
```diff
@@ -98,8 +98,7 @@ USE_MSGPACK_OPT=yes
 DISABLE_JEMALLOC?=no
 DISABLE_SYSC?=no
 DISABLE_CISCAT?=no
-IMAGE_TRUST_CHECKS?=1
-CA_NAME?=DigiCert High Assurance EV Root CA
```
[the remark this answers is not in the document]
Comment (tdrauncieal) [reply src/Makefile]: was do it!

File: src/Makefile [hunk 7]
```diff
@@ -622,29 +621,28 @@ failtarget:
...
+	@echo "   make DISABLE_JEMALLOC=yes     Not to build the JEMalloc library. Allowed values are 1, yes, YES, y, and Y, otherwise, the flag is ignored"
+	@echo "   make OFLAGS=-Ox              Overrides optimization level"
+	@echo "   make DISABLE_SYSC=yes        Not to build the Syscollector module (for unsupported systems). Allowed values are 1, yes, YES, y and Y, otherwise, the flag is ignored"
+	@echo "   make DISABLE_CISCAT=yes      Not to build the CIS-CAT module (for unsupported systems). Allowed values are 1, yes, YES, y and Y, otherwise, the flag is ignored"
+	@echo "   make OPTIMIZE_CPYTHON=yes    Enable this flag to optimize the python interpreter build, which is performed when used PYTHON_SOURCE. Allowed values are 1, yes, YES, y and Y, otherwise, the flag is ignored"
+	@echo "   make PYTHON_SOURCE=yes       Used along the deps target. Downloads the sources needed to build the python interpreter. Allowed values are 1, yes, YES, y and Y, otherwise, the flag is ignored"
+	@echo "   make HTTP_REQUEST_BRANCH=main Build http-request tool from selected branch, Allowed values are any existing branch.
```
[the remark this answers is not in the document]
Comment (tdrauncieal) [reply src/Makefile]: was fixed, it recovered

File: src/data_provider/src/packages/packageLinuxParserHelper.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (tdrauncieal) [reply src/data_provider/src/packages/packageLinuxParserHelper.h:160]: > I think more than updating the comment, the code must be changed to implement the feature. We can not accept a package without a name or version.

Please, explain about this, in case of missing name and version values, should it return empty json? e.g.

File: src/data_provider/CMakeLists.txt [hunk 8]
```diff
@@ -159,8 +163,9 @@ elseif(APPLE)
   target_link_libraries(sysinfo cjson ${SRC_FOLDER}/external/libplist/bin/lib/libplist-2.0.a)
 endif(CMAKE_SYSTEM_NAME STREQUAL "Windows")
 
-target_link_libraries(sysinfo wazuhext)
+target_link_libraries(sysinfo wazuhext urlrequest)
 
+add_subdirectory(${SRC_FOLDER}/shared/http-request ${SRC_FOLDER}/shared/http-request/build)
```
Comment (pereyra-m) [inline src/data_provider/CMakeLists.txt:168]: Shouldn't we use `add_subdirectory()` only for the folders we expect to contain a CMakeLists file?

File: src/data_provider/CMakeLists.txt [hunk 9]
```diff
@@ -159,8 +163,9 @@ elseif(APPLE)
   target_link_libraries(sysinfo cjson ${SRC_FOLDER}/external/libplist/bin/lib/libplist-2.0.a)
 endif(CMAKE_SYSTEM_NAME STREQUAL "Windows")
 
-target_link_libraries(sysinfo wazuhext)
+target_link_libraries(sysinfo wazuhext urlrequest)
```
Comment (pereyra-m) [inline src/data_provider/CMakeLists.txt:166]: If now the `sysinfo` library requires `urlrequest`, we should either link this dependency statically or distribute it with the installer. Please make and end-to-end test with an installed agent by sources and packages.

File: src/data_provider/src/sharedDefs.h [hunk 10]
```diff
@@ -21,17 +21,11 @@ constexpr auto WM_SYS_IF_FILE {"/etc/network/interfaces"};
...
 constexpr auto WM_SYS_NET_DIR {"/proc/net/" };
-
 constexpr auto DPKG_PATH {"/var/lib/dpkg/"};
 constexpr auto DPKG_STATUS_PATH {"/var/lib/dpkg/status"};
-
 constexpr auto RPM_PATH {"/var/lib/rpm/"};
-
 constexpr auto PACMAN_PATH {"/var/lib/pacman"};
-
-constexpr auto APK_PATH {"/lib/apk/db"};
```
Comment (pereyra-m) [inline src/data_provider/src/sharedDefs.h:32]: We shouldn't remove this.

File: src/data_provider/src/packages/packageLinuxDataRetriever.h [hunk 11]
```diff
@@ -88,10 +87,11 @@ class FactoryPackagesCreator<LinuxType::STANDARD> final
                 getRpmInfo(callback);
             }
 
-            if (Utils::existsDir(APK_PATH))
+            if (Utils::existsDir(SNAP_PATH))
```
Comment (pereyra-m) [inline src/data_provider/src/packages/packageLinuxDataRetriever.h:90]: I think we shouldn't replace the `APK_PATH` implementation.

**Review by pereyra-m — CHANGES_REQUESTED** [review]
- There are compilation errors in the checks, please verify them. It might be related to the GCC versions that Jenkins is using. For example:
https://github.com/wazuh/wazuh/pull/15740/checks?check_run_id=12764813024
```
build_wazuh/agent/wazuh-agent-4.6.0/src/shared/http-request/src/urlRequest.hpp:42:83: error: wrong number of template arguments (2, should be 1)
   42 | class cURLRequest : public Utils::Builder<T, std::shared_ptr<IRequestImplementator>>
      |                                                                                   ^~
```

- There is an inconsistency in the CMakeLists minimum required version:
The **http-request** repository has a `cmake_minimum_required(VERSION 3.16.3)` and the **data_provider** has a `cmake_minimum_required(VERSION 3.12.4)`. 
See for example
https://github.com/wazuh/wazuh/pull/15740/checks?check_run_id=12764815607
```
CMake Error at /build_wazuh/rpmbuild/BUILD/wazuh-agent-4.6.0/src/shared/http-request/CMakeLists.txt:1 (cmake_minimum_required):
  CMake 3.16.3 or higher is required.  You are running version 3.12.4
```

- There is an inconsistency in the DEPS version and precompiled OS target: the **http-request** has some harcoded values that might not match with the version and/or OS of the current build
```
if(NOT EXTERNAL_DEPS_VERSION)
    set(EXTERNAL_DEPS_VERSION "19")
endif(NOT EXTERNAL_DEPS_VERSION)

if(NOT RESOURCES_URL)
    set(RESOURCES_URL "https://packages.wazuh.com/deps/${EXTERNAL_DEPS_VERSION}/libraries/sources")
endif(NOT RESOURCES_URL)

if(NOT PRECOMPILED_RESOURCES_URL)
    set(PRECOMPILED_RESOURCES_URL "https://packages.wazuh.com/deps/${EXTERNAL_DEPS_VERSION}/libraries/linux/amd64")
endif(NOT PRECOMPILED_RESOURCES_URL)
``` 

- The build in Solaris is failing, maybe there is a problem in the `make deps` stage because the folder **http-request** is empty

```
" solaris10_gcc9: The source directory",
" solaris10_gcc9: ",
" solaris10_gcc9: /export/home/vagrant/build/solaris10/repository/src/shared/http-request",
" solaris10_gcc9: ",
" solaris10_gcc9: does not contain a CMakeLists.txt file.",
" solaris10_gcc9: ",
```

- Please, check the `maxdepth` option in Solaris 11
 https://github.com/wazuh/wazuh/pull/15740/checks?check_run_id=12764845716
```
" solaris11_gcc9: find: bad option -maxdepth",
```

- The `shared` folder has a different meaning in this repository. Please consider adding the submodule to the `external` folder

**Comment by tdrauncieal** [comment]
> * There are compilation errors in the checks, please verify them. It might be related to the GCC versions that Jenkins is using. For example:
>   https://github.com/wazuh/wazuh/pull/15740/checks?check_run_id=12764813024
> 
> ```
> build_wazuh/agent/wazuh-agent-4.6.0/src/shared/http-request/src/urlRequest.hpp:42:83: error: wrong number of template arguments (2, should be 1)
>    42 | class cURLRequest : public Utils::Builder<T, std::shared_ptr<IRequestImplementator>>
>       |                                                                                   ^~
> ```
this behavior is related to duplicated files with different impmlementation (builder.hpp) between this project and `http-request` https://github.com/wazuh/wazuh-http-request/pull/26#discussion_r1170426295, anyways I do not know why the mentioned change doesn't work at Debian target check 
> 
> * There is an inconsistency in the CMakeLists minimum required version:
>   The **http-request** repository has a `cmake_minimum_required(VERSION 3.16.3)` and the **data_provider** has a `cmake_minimum_required(VERSION 3.12.4)`.
>   See for example
>   https://github.com/wazuh/wazuh/pull/15740/checks?check_run_id=12764815607
> 
> ```
I think that CMake version should be updated in `centos5 builder` and this the constraint should be updated in `data_provider`, 
> CMake Error at /build_wazuh/rpmbuild/BUILD/wazuh-agent-4.6.0/src/shared/http-request/CMakeLists.txt:1 (cmake_minimum_required):
>   CMake 3.16.3 or higher is required.  You are running version 3.12.4
> ```
> 
> * There is an inconsistency in the DEPS version and precompiled OS target: the **http-request** has some harcoded values that might not match with the version and/or OS of the current build
> 
> ```
> if(NOT EXTERNAL_DEPS_VERSION)
>     set(EXTERNAL_DEPS_VERSION "19")
> endif(NOT EXTERNAL_DEPS_VERSION)
> 
> if(NOT RESOURCES_URL)
>     set(RESOURCES_URL "https://packages.wazuh.com/deps/${EXTERNAL_DEPS_VERSION}/libraries/sources")
> endif(NOT RESOURCES_URL)
> 
> if(NOT PRECOMPILED_RESOURCES_URL)
>     set(PRECOMPILED_RESOURCES_URL "https://packages.wazuh.com/deps/${EXTERNAL_DEPS_VERSION}/libraries/linux/amd64")
> endif(NOT PRECOMPILED_RESOURCES_URL)
> ```
I see a deafult value setting here
> 
> * The build in Solaris is failing, maybe there is a problem in the `make deps` stage because the folder **http-request** is empty
> 
> ```
> " solaris10_gcc9: The source directory",
> " solaris10_gcc9: ",
> " solaris10_gcc9: /export/home/vagrant/build/solaris10/repository/src/shared/http-request",
> " solaris10_gcc9: ",
> " solaris10_gcc9: does not contain a CMakeLists.txt file.",
> " solaris10_gcc9: ",
> ```
> 
> * Please, check the `maxdepth` option in Solaris 11
>   https://github.com/wazuh/wazuh/pull/15740/checks?check_run_id=12764845716
> 
> ```
> " solaris11_gcc9: find: bad option -maxdepth",
> ```
> 
> * The `shared` folder has a different meaning in this repository. Please consider adding the submodule to the `external` folder
These errors are related to each other, because the version of `find tool` in `solaris11` does not support this modifier, I'll work on this

File: src/data_provider/src/packages/packageLinuxDataRetriever.h (hunk 11 above)
Comment (tdrauncieal) [reply src/data_provider/src/packages/packageLinuxDataRetriever.h:90]: Disregard this change made because the l function was replaced
image

File: src/data_provider/src/sharedDefs.h (hunk 10 above)
Comment (tdrauncieal) [reply src/data_provider/src/sharedDefs.h:32]: Same of next comment answered I understand what implementation was changed

File: src/data_provider/CMakeLists.txt (hunk 8 above)
Comment (tdrauncieal) [reply src/data_provider/CMakeLists.txt:168]: Sure, but I wonder why it doesn't fail here?

File: src/data_provider/src/packages/packageLinuxDataRetriever.h (hunk 11 above)
Comment (tdrauncieal) [reply src/data_provider/src/packages/packageLinuxDataRetriever.h:90]: I understood that this was no longer necessary, maybe we should review this change.

Comment (MiguelazoDS) [reply src/data_provider/src/packages/packageLinuxDataRetriever.h:90]: Why do you think this is no longer necessary? this is present in the master branch and it is part of the Alpine Linux package inventory support implementation.

File: src/data_provider/src/packages/packageLinuxDataRetriever.h [hunk 12]
```diff
@@ -88,10 +87,11 @@ class FactoryPackagesCreator<LinuxType::STANDARD> final
...
             {
-                getApkInfo(APK_DB_PATH, callback);
```
Comment (MiguelazoDS) [inline src/data_provider/src/packages/packageLinuxDataRetriever.h:93]: Careful during rebases, here we are modifying other developments.

File: src/data_provider/src/sharedDefs.h (hunk 10 above)
Comment (MiguelazoDS) [reply src/data_provider/src/sharedDefs.h:32]: This is present in master, and shouldn't be modified here. 

image

**Review by MiguelazoDS — CHANGES_REQUESTED** [review]

File: src/data_provider/src/packages/packageLinuxDataRetriever.h (hunk 12 above)
Comment (tdrauncieal) [reply src/data_provider/src/packages/packageLinuxDataRetriever.h:93]: rebases with inactive projects for so long are always tricky, even using `-X theirs` switches, I'll check this out

File: src/data_provider/tests/sysInfoPackagesLinuxHelper/sysInfoPackagesLinuxHelper_test.cpp
Comment (MiguelazoDS) [inline src/data_provider/tests/sysInfoPackagesLinuxHelper/sysInfoPackagesLinuxHelper_test.cpp:1]: All the tests for Alpine Linux were overridden with this PR, please check that.

**Review by MiguelazoDS — CHANGES_REQUESTED** [review]

File: src/data_provider/src/packages/packageLinuxDataRetriever.h (hunk 12 above)
Comment (pereyra-m) [reply src/data_provider/src/packages/packageLinuxDataRetriever.h:93]: I'd avoid the use of `-X theirs` and resolve manually every conflict to inspect the code and avoid any silent code add/removal

Comment (MiguelazoDS) [reply src/data_provider/src/packages/packageLinuxDataRetriever.h:93]: I didn't know that option exist, but that is exactly the issue. When doing a rebase "theirs" is your implementation because you are in the branch to rebase against (master in this case). So it will always ignore what is master (in this case) branch and will replace its content with the new code you are adding.
I do what @pereyra-m does.  It gives you more control and I think is the way to go, but I can't agree more about rebasing old branches, it is always tedious.

File: src/data_provider/CMakeLists.txt (hunk 9 above)
Comment (tdrauncieal) [reply src/data_provider/CMakeLists.txt:166]: `http-request` is a submodule or can obtain it from make deps stage

File: src/data_provider/src/packages/packageLinuxDataRetriever.h (hunk 12 above)
Comment (tdrauncieal) [reply src/data_provider/src/packages/packageLinuxDataRetriever.h:93]: anyway, whether with `-X theirs` or not, checking with the base branch is the key

**Review by MiguelazoDS — COMMENTED** [review]
Isn't this suppose to work? The project does not compile.
image

**Comment by tdrauncieal** [comment]
> ¿No se supone que esto funciona? El proyecto no compila. imagen


this PR has a dependency as you can see in the description
I don't know in detail about git submodules but the first time you need `git pull --init --recurse-submodules`
in which case it will get the main branch and **also fail** .
But you can use the new functionality enabled to get this submodule **like dependency** as described on the cover, getting the `15429-add-support-to-get-packages-from-snap` branch, from the `http-request` project

**Comment by MiguelazoDS** [comment]
> > Isn't this suppose to work? The project does not compile.imagen
> 
> this PR has a dependency as you can see in the description I don't know in detail about git submodules but the first time you need `git pull --init --recurse-submodules` in which case it will get the `main` branch and **also fail** due the previously mentioned dependency. But you can use the new functionality enabled to get this submodule **like dependency** as described on the cover, getting the `15429-add-support-to-get-packages-from-snap` branch, from the `http-request` project

I don't consider a submodule to be a dependency in the way we consider external dependencies by using make deps. 

> `git pull --init --recurse-submodules` in which case it will get the `main` branch and **also fail** due the previously mentioned dependency

Maybe we are referring to different things, what is the dependency in this case? 

> But you can use the new functionality enabled to get this submodule **like dependency** as described on the cover.

I saw that, but I was not sure about the approach. It looks like a WA to get the http-request but I don't see how can we integrate this with the whole Wazuh project. I'll keep taking a look at this to better understand the implementation.

**Comment by tdrauncieal** [comment]
@MiguelazoDS 
This functionality was requested to aid the CICD team and in environments where it can be difficult to obtain this submodule.

**Comment by Dwordcito** [comment]
@tdrauncieal 
1 check the Solaris ci execution.
2 dependencies of http-request are downloaded and built twice, explore the possibility of setting a variable from data provider, to download and uncompress in wazuh/src/external. Currently is downloaded en wazuh/src/data_provider/external.

this change should be in wazuh/http-request as well.

**Comment by Dwordcito** [comment]
3 - Modify the documentation (add snap)
image

File: src/data_provider/src/packages/packageLinuxParserHelper.h [hunk 13]
```diff
@@ -133,6 +135,103 @@ namespace PackageLinuxHelper
...
+
+            if (publisher.contains("display-name"))
+            {
+                vendor = publisher.at("display-name");
+            }
+        }
+
+        if (info.contains("install-date"))
+        {
+            install_time = info.at("install-date");
```
Comment (Dwordcito) [inline src/data_provider/src/packages/packageLinuxParserHelper.h:189]: Currently format:
image

Expected format:
2018/02/08 18:45:48

**Review by Dwordcito — CHANGES_REQUESTED** [review]

File: src/data_provider/src/packages/packageLinuxParserHelper.h (hunk 13 above)
Comment (tdrauncieal) [reply src/data_provider/src/packages/packageLinuxParserHelper.h]: Fixed!

**Comment by tdrauncieal** [comment]
> 3 - Modify the documentation (add snap) image

fixed: https://github.com/wazuh/wazuh-documentation/pull/6084

**Review by matias-braida — APPROVED** [review]
GJ

**Review by matias-braida — APPROVED** [review]
LGTM

File: src/data_provider/src/packages/packageLinuxParserSnap.cpp [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+
+            if (!mapping.empty())
+            {
+                callback(mapping);
+            }
+        }
+    },
+    [&](const std::string & result, const long responseCode)
+    {
+        std::cerr << "Error retrieving packages using snap unix-socket (" << responseCode << ") " << result << "\n";
```
[the remark this answers is not in the document]
Comment (pereyra-m) [reply src/data_provider/src/packages/packageLinuxParserSnap.cpp:41]: Here we have an example of the daemon in foreground

**Review by pereyra-m — COMMENTED** [review]
Gj!

File: src/data_provider/CMakeLists.txt [hunk 15]
```diff
@@ -179,6 +191,13 @@ endif(CMAKE_SYSTEM_NAME STREQUAL "Windows")
 
 target_link_libraries(sysinfo wazuhext)
 
+if(CMAKE_SYSTEM_NAME STREQUAL "Linux")
+  target_link_libraries(sysinfo urlrequest)
+  if(CMAKE_CHECK_CENTOS5)
+      set(USE_HTTP 1)
+  endif(CMAKE_CHECK_CENTOS5)
+  add_subdirectory(${SRC_FOLDER}/shared_modules/http-request ${SRC_FOLDER}/shared_modules/http-request/build)
```
[the remark this answers is not in the document]
Comment (GabrielEValenzuela) [reply src/data_provider/CMakeLists.txt:199]: We need this due in case of absolute source path (and when this path isn't in your source tree), CMake cannot guess build directory, and you need to provide it explicitly.

File: .gitmodules
[the remark this answers is not in the document]
Comment (GabrielEValenzuela) [reply .gitmodules:1]: We'll keep this for future releases that will use `git submodules`

File: src/data_provider/src/packages/packageLinuxParserSnap.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (GabrielEValenzuela) [reply src/data_provider/src/packages/packageLinuxParserSnap.cpp:41]: For now, this will be the output until we have a unified logger, then the output will be refactor to the logger.

File: src/wazuh_db/schema_upgrade_v11.sql
[the remark this answers is not in the document]
Comment (GabrielEValenzuela) [reply src/wazuh_db/schema_upgrade_v11.sql:1]: Fix it. We add a new schema, the number 13

File: src/data_provider/CMakeLists.txt (hunk 15 above)
[the remark this answers is not in the document]
Comment (MiguelazoDS) [reply src/data_provider/CMakeLists.txt:199]: Isn't this a case for include_directories()?

File: src/data_provider/src/packages/packageLinuxParserSnap.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (pereyra-m) [reply src/data_provider/src/packages/packageLinuxParserSnap.cpp:41]: Ok, thank you!

File: src/data_provider/CMakeLists.txt (hunk 15 above)
[the remark this answers is not in the document]
Comment (pereyra-m) [reply src/data_provider/CMakeLists.txt:199]: Thank you for clarifying. 
So I understand now that you are specifying a binary dir

https://cmake.org/cmake/help/latest/command/add_subdirectory.html

> add_subdirectory(source_dir [binary_dir] [EXCLUDE_FROM_ALL] [SYSTEM])

File: .gitmodules
[the remark this answers is not in the document]
Comment (pereyra-m) [reply .gitmodules:1]: Thank you.

File: src/data_provider/src/packages/packageLinuxParserSnap.cpp [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+        HttpUnixSocketURL("/run/snapd.socket", "http://localhost/v2/snaps"),
+        [&](const std::string & result)
+    {
+        auto feed = nlohmann::json::parse(result).at("result");
+
+        int count = (int)feed.size();
+
+        for (int k = 0; k < count; k++)
+        {
+            auto& entry = feed.at(k);
```
[the remark this answers is not in the document]
Comment (GabrielEValenzuela) [reply src/data_provider/src/packages/packageLinuxParserSnap.cpp:29]: Fix it. Thanks ! =D

File: src/data_provider/CMakeLists.txt (hunk 15 above)
[the remark this answers is not in the document]
Comment (GabrielEValenzuela) [reply src/data_provider/CMakeLists.txt:199]: I think is not possible @MiguelazoDS , because the function `include_directories()` is used to add the given directories to those the compiler uses to search for include files.

https://cmake.org/cmake/help/latest/command/include_directories.html

File: src/data_provider/src/packages/packageLinuxParserHelper.h [hunk 17]
```diff
@@ -15,10 +15,12 @@
 #include "stringHelper.h"
 #include "json.hpp"
 #include "timeHelper.h"
+#include "sharedDefs.h"
 
 #pragma GCC diagnostic push
 #pragma GCC diagnostic ignored "-Wunused-function"
 
+#include <iostream>
```
[the remark this answers is not in the document]
Comment (GabrielEValenzuela) [reply src/data_provider/src/packages/packageLinuxParserHelper.h:23]: Fix it =D

File: src/data_provider/CMakeLists.txt [hunk 18]
```diff
@@ -68,7 +77,10 @@ link_directories(${SRC_FOLDER}/external/libarchive/.libs/)
 link_directories(${SRC_FOLDER}/external/rpm/builddir/)
 endif(NOT CMAKE_CHECK_CENTOS5)
 link_directories(${SRC_FOLDER}/external/openssl/)
-
+if(CMAKE_SYSTEM_NAME STREQUAL "Linux")
+  link_directories(${SRC_FOLDER}/shared_modules/http-request/build)
+endif(CMAKE_SYSTEM_NAME STREQUAL "Linux")
+  link_directories(${SRC_FOLDER}/external/curl/lib/.libs/)
```
[the remark this answers is not in the document]
Comment (GabrielEValenzuela) [reply src/data_provider/CMakeLists.txt:81]: Fix it =D

File: src/data_provider/tests/sysInfoPackagesLinuxHelper/sysInfoPackagesLinuxHelper_test.cpp
[the remark this answers is not in the document]
Comment (GabrielEValenzuela) [reply src/data_provider/tests/sysInfoPackagesLinuxHelper/sysInfoPackagesLinuxHelper_test.cpp:1]: New test case added

File: src/data_provider/src/packages/packageLinuxParserSnap.cpp [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+    {
+        auto feed = nlohmann::json::parse(result, nullptr, false).at("result");
+
+        if (feed.is_discarded())
+        {
+            std::cerr << "Error parsing JSON feed\n";
+        }
+
+
+        int count = (int)feed.size();
```
Comment (pereyra-m) [inline src/data_provider/src/packages/packageLinuxParserSnap.cpp:31]: This line is unused

File: src/wazuh_db/schema_upgrade_v13.sql [hunk 20]
```diff
@@ -0,0 +1,44 @@
...
+    msu_name TEXT,
+    checksum TEXT NOT NULL CHECK (checksum <> ''),
+    item_id TEXT,
+    PRIMARY KEY (scan_id, name, version, architecture, location)
+);
+
+INSERT INTO _sys_programs SELECT scan_id, scan_time, format, name, priority, section, size, vendor, install_time, version, architecture, multiarch, source, description, location, triaged, cpe, msu_name, CASE WHEN checksum <> '' THEN checksum ELSE 'legacy' END AS checksum, item_id FROM sys_programs;
+DROP TABLE IF EXISTS sys_programs;
+ALTER TABLE _sys_programs RENAME TO sys_programs;
+CREATE INDEX IF NOT EXISTS programs_id ON sys_programs (scan_id);
```
Comment (pereyra-m) [inline src/wazuh_db/schema_upgrade_v13.sql:40]: We should also recreate the trigger of this table.

**Review by pereyra-m — COMMENTED** [review]
Gj!

Please, remember to avoid general commit messages like "code review changes". 
Let's be explicit instead.

**Review by pereyra-m — DISMISSED** [review]
LGTM!

**Review by MiguelazoDS — DISMISSED** [review]
GJ!
Agent Jammy
<details><summary>Show image</summary>



</details>

Manager Focal

<details><summary>Show image</summary>



</details>

**Review by Dwordcito — APPROVED** [review]
